capsules/src/apds9960.rs

```diff
@@ -0,0 +1,400 @@
+    TakeMeasurement3,
+    TakeMeasurement4,
+
+    /// States for optional chip functionality
+    SetPulse, // Set proximity pulse
+    SetLdrive, // Set LED Current for Prox and ALS sensors
+    Done,      // Final state for take_measurement() state sequence
+}
+
+pub struct APDS9960<'a> {
+    i2c: &'a dyn i2c::I2CDevice,
+    interrupt_pin: &'a dyn gpio::InterruptPin,
+    prox_callback: OptionalCell<&'a dyn kernel::hil::sensors::ProximityClient>,
+    state: Cell<State>,
+    buffer: TakeCell<'static, [u8]>,
+}
+
+impl<'a> APDS9960<'a> {
+    pub fn new(
+        i2c: &'a dyn i2c::I2CDevice,
+        interrupt_pin: &'a dyn gpio::InterruptPin,
+        buffer: &'static mut [u8],
+    ) -> APDS9960<'a> {
+        // setup and return struct
+        APDS9960 {
+            i2c: i2c,
+            interrupt_pin: interrupt_pin,
+            prox_callback: OptionalCell::empty(),
+            state: Cell::new(State::Idle),
+            buffer: TakeCell::new(buffer),
+        }
+    }
+
+    // Read I2C-based ID of device (should be 0xAB)
+    pub fn read_id(&self) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            buffer[0] = Registers::ID as u8;
+            self.i2c.write_read(buffer, 1, 1);
+
+            self.state.set(State::ReadId); // Reading ID
+        });
+    }
+
+    // Set Proximity Pulse Count and Length(1 = default)
+    pub fn set_proximity_pulse(&self, mut length: u8, mut count: u8) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            if length > 3 {
+                length = 3;
+            }
+            if count > 63 {
+                count = 63;
+            }
+
+            buffer[0] = Registers::PROXPULSEREG as u8;
+            buffer[1] = (length << 6 | count) as u8;
+            self.i2c.write(buffer, 2);
+
+            self.state.set(State::SetPulse); // Send pulse control command to device
+        });
+    }
+
+    // Set LED Current Strength (0 -> 100 mA , 3 --> 12.5 mA)
+    pub fn set_ldrive(&self, mut ldrive: u8) {
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            if ldrive > 3 {
+                ldrive = 3;
+            }
+
+            buffer[0] = Registers::CONTROLREG1 as u8;
+            buffer[1] = (ldrive << 6) as u8;
+            self.i2c.write(buffer, 2);
+
+            self.state.set(State::SetLdrive); // Send LED Current Control gain
+        });
+    }
+
+    // Take measurement immediately
+    pub fn take_measurement(&self) {
+        // Enable power and proximity sensor
+        self.buffer.take().map(|buffer| {
+            self.i2c.enable();
+
+            buffer[0] = Registers::ENABLE as u8;
+            buffer[1] = PON | PEN;
+
+            self.i2c.write(buffer, 2);
+
+            self.state.set(State::TakeMeasurement1);
+        });
+    }
+
+    // Take Simple proximity measurement with persistence=4, low threshold=0, high threshold=175, and Sleep-After-Interrupt Mode enabled
```

Comment:
this comment seems out of date, as the thresholds are now inputs to the function

Reply:
Done.